lib/runtime-core/src/memory/mod.rs

```diff
@@ -0,0 +1,240 @@
+use crate::{
+    export::Export,
+    import::IsExport,
+    memory::dynamic::DYNAMIC_GUARD_SIZE,
+    memory::static_::{SAFE_STATIC_GUARD_SIZE, SAFE_STATIC_HEAP_SIZE},
+    types::{MemoryDesc, ValueType},
+    vm,
+};
+use std::{cell::RefCell, fmt, mem, ptr, rc::Rc, slice};
+
+pub use self::dynamic::DynamicMemory;
+pub use self::static_::{SharedStaticMemory, StaticMemory};
+
+mod dynamic;
+mod static_;
+
+pub const WASM_PAGE_SIZE: usize = 65_536;
+pub const WASM_MAX_PAGES: usize = 65_536;
+
+pub struct Memory {
+    desc: MemoryDesc,
+    storage: Rc<RefCell<(MemoryStorage, Box<vm::LocalMemory>)>>,
+}
+
+impl Memory {
+    pub fn new(desc: MemoryDesc) -> Option<Self> {
+        let mut vm_local_memory = Box::new(vm::LocalMemory {
+            base: ptr::null_mut(),
+            bound: 0,
+            memory: ptr::null_mut(),
+        });
+
+        let memory_storage = match desc.memory_type() {
+            MemoryType::Dynamic => {
+                MemoryStorage::Dynamic(DynamicMemory::new(desc, &mut vm_local_memory)?)
+            }
+            MemoryType::Static => {
+                MemoryStorage::Static(StaticMemory::new(desc, &mut vm_local_memory)?)
+            }
+            MemoryType::SharedStatic => unimplemented!("shared memories are not yet implemented"),
+        };
+
+        Some(Memory {
+            desc,
+            storage: Rc::new(RefCell::new((memory_storage, vm_local_memory))),
+        })
+    }
+
+    pub fn description(&self) -> MemoryDesc {
```

Comment:
Should this be `descriptor` E.g. `tableDescriptor`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Memory#Syntax

Reply:
Hmm, do you have a preference to use the `descriptor` term rather than `description`?

Reply:
Yes, I think `descriptor` is the common terminology that I've seen in the `js-api`